src/storage/sql_table.cpp

```diff
@@ -19,64 +43,329 @@ SqlTable::SqlTable(BlockStore *const store, const catalog::Schema &schema, const
+    auto old_dt_version = tables_.Find(old_version)->second;
+
+    // The slot version is not the same as the version_num
+    col_id_t original_column_ids[redo->NumColumns()];
+    ModifyProjectionHeaderForVersion(redo, tables_.Find(version_num)->second, old_dt_version, original_column_ids);
+
+    // Get the result and copy back the old header
+    bool result = old_dt_version.data_table->Update(txn, slot, *redo);
+
+    // We should create a buffer of old Projected Row and update in place. We can't just
+    // directly erase the data without creating a redo and update the chain.
+
+    // auto old_pair = InitializerForProjectedRow(redo_col_oids, old_version);
+
+    // // 1. Create a ProjectedRow Buffer for the old version
+    // byte *buffer = common::AllocationUtil::AllocateAligned(old_pair.first.ProjectedRowSize());
+    // storage::ProjectedRow *pr = old_pair.first.InitializeRow(buffer);
+
+    // // 2. Copy from new ProjectedRow to old ProjectedRow
+    // StorageUtil::CopyProjectionIntoProjection(redo, map, tables_.Find(old_version)->second.layout, pr,
+    // old_pair.second);
+
+    // // 3. Update the old data-table
+    // bool result = tables_.Find(old_version)->second.data_table->Update(txn, slot, *pr);
+    // delete[] buffer;
+    if (!result) {
+      return {false, slot};
+    }
+    ret_slot = slot;
+  } else {
+    STORAGE_LOG_DEBUG("have to delete and insert ... ");
+
+    // need to create a new ProjectedRow of all columns
+    // 1. Get the old row
+    // 2. Convert it into new row
+    // 3. Delete old row
+    // 4. Update the new row before insert
+    // 5. Insert new row into new table
+
+    // 1. Get old row
+    // 2. Convert it into new row
+    std::vector<catalog::col_oid_t> new_col_oids;  // the set of col oids which the new schema has
+    for (auto &it : tables_.Find(version_num)->second.column_map) new_col_oids.emplace_back(it.first);
+    auto new_pair = InitializerForProjectedRow(new_col_oids, version_num);
+    auto new_buffer = common::AllocationUtil::AllocateAligned(new_pair.first.ProjectedRowSize());
+    ProjectedRow *new_pr = new_pair.first.InitializeRow(new_buffer);
+    bool valid = Select(txn, slot, new_pr, new_pair.second, version_num);
+    if (!valid) {
+      delete[] new_buffer;
+      return {false, slot};
+    }
+    // 3. Delete the old row
+    bool succ = tables_.Find(old_version)->second.data_table->Delete(txn, slot);
+
+    // 4. Update the new row before insert
+    StorageUtil::CopyProjectionIntoProjection(*redo, map, tables_.Find(version_num)->second.layout, new_pr,
+                                              new_pair.second);
+
+    // 5. Insert the row into new table
+    storage::TupleSlot new_slot;
+    if (succ) {
+      new_slot = tables_.Find(version_num)->second.data_table->Insert(txn, *new_pr);
+    } else {
+      // someone else deleted the old row, write-write conflict
+      delete[] new_buffer;
+      return {false, slot};
+    }
+
```

Comment:
I think there might be a logic problem here regarding varlen contents.
For the new row, you only did a shallow copy. For varlen contents, if you deleted the old-version row, GC can eventually reclaim all its contents. Then for the updated row, its varlen_content will point to already freed memory.